resolver/resolver.go

```diff
@@ -20,8 +31,82 @@ var (
 	recurseCnt = 3
 )
 
-// resolveOut queries other nameserver
-// randomly picks from the list that is not mesos
+// holds configuration state and the resource records
+type Resolver struct {
+	Version    string
+	Config     records.Config
+	rs         records.RecordGenerator
+	leader     string
+	leaderLock sync.RWMutex
+}
+
+// launches DNS server for a resolver
+func (res *Resolver) LaunchDNS() {
+	// Handers for Mesos requests
+	dns.HandleFunc(res.Config.Domain+".", panicRecover(res.HandleMesos))
+	// Handler for nonMesos requests
+	dns.HandleFunc(".", panicRecover(res.HandleNonMesos))
+
+	go res.Serve("tcp")
+	go res.Serve("udp")
+}
+
+// starts a DNS server for net protocol (tcp/udp)
+func (res *Resolver) Serve(net string) {
+	defer func() {
+		if rec := recover(); rec != nil {
+			logging.Error.Printf("%s\n", rec)
+			os.Exit(1)
+		}
+	}()
+
+	server := &dns.Server{
+		Addr:       res.Config.Listener + ":" + strconv.Itoa(res.Config.Port),
+		Net:        net,
+		TsigSecret: nil,
+	}
+
+	err := server.ListenAndServe()
+	if err != nil {
+		logging.Error.Printf("Failed to setup "+net+" server: %s\n", err.Error())
+	} else {
+		logging.Error.Printf("Not listening/serving any more requests.")
+	}
+
+	os.Exit(1)
+}
+
+// launches Zookeeper detector
+func (res *Resolver) LaunchZK() {
+	dr, err := res.ZKdetect()
+	if err != nil {
+		logging.Error.Println(err.Error())
+		os.Exit(1)
+	}
+
+	logging.VeryVerbose.Println("Warning: waiting for initial information from Zookeper.")
+	select {
+	case <-dr:
+		logging.VeryVerbose.Println("Warning: got initial information from Zookeper.")
+	case <-time.After(4 * time.Minute):
+		logging.Error.Println("timed out waiting for initial ZK detection, exiting")
+		os.Exit(1)
+	}
+}
+
+// triggers a new refresh from mesos master
+func (res *Resolver) Reload() {
+	t := records.RecordGenerator{}
+	err := t.ParseState(res.leader, res.Config)
+
+	if err == nil {
+		res.rs = t
```

Comment:
access to res.rs is not safe. it's read from and written to from different goroutines.